_episodes/07-b-numpy.md

```diff
@@ -0,0 +1,280 @@
+---
+title: "Features of Numpy Arrays"
+teaching: 40
+exercises: 20
+questions:
+- "What are the differences between numpy arrays and lists?"
+- "How can I use NumPy to do calculations?"
+objectives:
+- "Be able to name the differences between Python lists and numpy arrays."
+- "Understand the idea of broadcasting"
+keypoints:
+- "NumPy arrays which are the same size use element-wise operations when added or subtracted"
+- "NumPy uses something called *broadcasting* for arrays which are not the same size to allow arrays to be added or multiplied."
+- "NumPy has extensive documentation online - you "
+---
+
+[Numpy](https://numpy.org/) is a widely used Python library for scientific computing. It has a number of useful features, including the a data structure called an array which we have already worked with a bit.
+
+## NumPy Arrays vs. Python Lists
+
+Previously, you have worked with the built-in types of `lists`. NumPy arrays seem similar, but offer some distinct advantages. We have also worked with Numpy arrays, but we have not taken advantage of their unique features. 
+
+Numpy arrays take up less space, are faster, and have more mathematical operations associated with them. However, unlike lists, they elements all have to be the same type. This is why we had to read in the file from the Tabular Data lesson using the `dtype` argument.
+
+There are also differences in how lists and numpy arrays behave. Let's look at some of these.
+
+To use the numpy library, we have to import it. When `numpy` is imported, it is often shortened to `np`.
+
+~~~
+import numpy as np
+
+# Create some lists
+r1 = [0.0, 0.1, 0.0]
+r2 = [0.0, 0.0, 0.0]
+
+# Create some numpy arrays
+r1_array = np.array(r1)
+r2_array = np.array(r2)
+~~~
+{: .language-python}
+
+In this code block, we've created two lists (`r1` and `r2`). We then created numpy array versions of these lists.
+
+When we print `r1` and `r1_array`, the results look very similar.
+
+~~~
+print(r1)
+print(r1_array)
+~~~
+{: .language-python}
+
+~~~
+[0.0, 0.1, 0.0]
+[0.  0.1 0. ]
+~~~
+{: .output}
+
+However, one way numpy arrays and lists vary significantly is that you can easily perform element-wise operations on numpy arrays without loops. **This is advantageous because loops in Python are slow. However, `numpy` functions are optimized to run quickly.** If possible, you should avoid loops when working with numpy arrays.
```

Comment:
I would say "are different" instead of "vary significantly".  Maybe we can soften the language on "avoid loops" since we exactly taught them how to use loops with numpy arrays in the previous lesson.  Maybe something like "You can make your code much faster if you use numpy element-by-element operations instead of loops if all of your data is the same type"